### ISSUE TITLE
proposal: FiniteStream

### ISSUE THREAD

**Issue by tatumizer** [dart-lang/language#224]
Not sure whether the issue belongs here or in SDK, but it's closely related to the language construct "await for", so for now posting it here.
"await for" is inherently **unsafe**. If applied to an infinite stream, it never ends, and the program "blocks" forever. In some contexts, this behavior probably can be tolerated (but those are exactly the cases where "stream.listen()" would be a much better fit). For most users, infinitely long hanging would be surprising.

So the idea is: 
- introduce a class `FiniteStream(Stream stream, Duration timeoutTillDone)`. TimeoutTillDone is a mandatory parameter.
- For FiniteStream, define an *async* method `toIterable()` (we can't define a constructor Iterable.fromFiniteStream b/c constructors cannot be async)
- in "await for", allow iterations only over finite streams. Though as soon as we have "toIterable()" (see above), the case for having an "await for" operator in the language becomes quite unconvincing... but there's nothing can be done about it at this point.

**Comment by munificent** [comment]
> * in "await for", allow iterations only over finite streams.

This would be a breaking change since users can write code now that awaits non-finite streams.

I don't think this is a real concern in practice. Users can just as easily write `for (;;) ...` or `while (true) ...` and that doesn't seem to cause significant problems. Dart is Turing complete. :)

**Comment by tatumizer** [comment]
it's a logical counter-argument, and I don't have a strong opinion, but if people really use it, isn't it strange that no one requested a method for setting a "timeout till done" on stream (stream has another timeout - for a time interval between events, not for the whole operation).

**Comment by munificent** [comment]
I think people do request timeout control for streams, and I think there's various ways to do that, though I don't know the details. (I honestly don't use streams much and find them pretty confusing.) I think it's something that can get solved pretty well at the library level.

**Comment by tatumizer** [comment]
I will try to open request against SDK then, asking for another kind of stream timeout, let's see what they have to say :) Thanks!

**Comment by pulyaevskiy** [comment]
> isn't it strange that no one requested a method for setting a "timeout till done" on stream

I think a reason could be because there is no clear definition of "done" in this case and it is normally something you can only tell from the context of your problem.

From a `Stream` point of view, it is considered done when you call `streamController.close()`, it's up to the creator of that stream to decide when to call this method. There is no difference between `await for` and `stream.listen()` in this regard.

**Comment by tatumizer** [comment]
The difference is that with "listen", nobody waits until the stream is "done", whatever that means.
With "await for" (citing from documentation)

> After each iteration, the function enclosing the loop suspends until the next element is available or the stream is **done**.

Any function containing "await for" has to return some Future (it has to be async function after all).
Before the stream is "done", the caller function awaiting for this Future to complete will be blocked.
https://stackoverflow.com/questions/42611880/difference-between-await-for-and-listen-in-dart

It's not me who is saying "done", it's the documentation :)
Streams are confusing enough as they are, but the features like this are just mind-blowing :)

**Comment by pulyaevskiy** [comment]
I guess for me this behavior seems logical. This is consistent with how regular for-loop works, i.e. it blocks execution of a function until it iterates over all elements of a collection. Basically the same as what @munificent pointed out, in case of an infinite stream we get a `while (true)` scenario.

**Comment by tatumizer** [comment]
In while(true) loop, it's a good idea to put a limit on a number of iterations, just in case. (Unless it's absolutely obvious from the code that the loop ends). In sync programming, the number of iterations correlate with the timing, so you know that in case of a bug, it won't cycle longer than, say, 1 minute, and then you raise an Exception.

With "await for", this doesn't work, so to be defensive, you will have to implement this timeout yourself.
The whole issue was to standardize on this timeout. After all, the Future has a timeout - you could also question the rationale behind this timeout, based on the same arguments.
And why implement any defensive features at all? All these non-nullables, immutable types, and even types in general to begin with? Dart 1 had optional types, and there were also philosophical arguments explaining that of course it doesn't need (strong) types and would be much better off without them.

Not sure how many people use "await for" though,  maybe no one, Then it's a purely academic discussion, kind of a tempest in a (empty?) teapot.

**Comment by zoechi** [comment]
I think all that is needed is a stream transformer that takes a stream and a timeout.
The transformer forwards all events until timeout is reached and then closes the output stream.

**Comment by tatumizer** [comment]
> The transformer forwards all events until timeout is reached and then closes the output stream.
timeout is an error condition, almost by definition, so it should throw, not silently close the stream. And the syntax should be much shorter because at least in a context of "await for", timeout should be *common*, and "no timeout" must be rare.
Speaking by analogy with while(true) - sure, you can leave it unchecked, but after having a couple of embarrassments with infinite loops in your life... A burnt child dreads the fire :) (There's a funnier version of this proverb in Russian: "ожегшись на молоке, дуют на воду", but google translate can't really translate it).

**Comment by zoechi** [comment]
> Timeout is an error condition

You can make a transformer that emits an error as well.
It just depends on what you want. 

> the syntax should be much shorter

I hope extension methods will finally make stream transformers more convenient to use

**Comment by lrhn** [comment]
Infinite behavior is not unique to streams. Iterables can also be infinite, and being finite is not a guarantee of not *effectively* never ending: `var lotsanums = Iterable<int>.generate(1<<62);` (iterating that would take >200 days on my computer).

There is no meaningful distinction between a finite stream and a very long stream.
There is no difference at all between a stream controller for a finite stream and a *potentially* infinite stream, we cannot force you to close the finite stream anyway. The only way a "finite stream" could make sense if you limit the length *up front*.
That still doesn't account for an incomplete stream which never emits an event again, which will also hang your program if you wait for it.

Potentially infinite behavior is everywhere in programming and we have learned to live with it, mainly because the Halting Problem is undecidable.

Whether you do 
```dart
await for (var e in stream) { ... }
``` 
or 
```dart
for (var e in await stream.toList()) { ... }
```
or
```dart
for (var q = StreamQueue(stream); await q.hasNext(); ) { var e = await q.next(); ... }
``` 
doesn't matter, either you introduce a limit yourself, or you run until the source decides to stop. Or forever if it doesn't. Only you can know if the limit is reasonable.


Also: [`Stream.timeout`](https://api.dartlang.org/stable/2.1.1/dart-async/Stream/timeout.html). It adds the timeout *between* events, so you can stop if the stream becomes inactive. Using a timeout to stop events seems overkill and unpredictable, I'd rather use `.take(1000000)` for limiting infinite behavior. I can do an `OutOfMemory` in very little *time*.

**Comment by tatumizer** [comment]
@lrhn: I opened the issue because I couldn't make sense of `await for` in the context of literals. I agree with your arguments (overall time limit doesn't do much good). Also, I overlooked stream.toList().
However, you provided me with the ammunition to strengthen the case against "for-element" and "await-for-element" for literals.
Earlier, we agreed to frown upon having exactly 2 ways to express exactly the same thing.
Now, it turns out that `[...iterable]` means the same as `[for (x in iterable) x]`, and `[...stream.toList()]` - same thing as `[await for (x in stream) x]`. Do you see much point in supporting both forms?

**Comment by lrhn** [comment]
@tatumizer 
Yes.
I am not opposed to having two ways to do the same thing, as long as one of them is a specialized abstraction for a particular common use-case, and makes that use-case easier to read and write, and it's clear when you should use the optimized form.

We allow `"$x"` and `"${x}"` as string interpolations, but the former occurs often enough that it's worth it.
We allow `for (var x in something) {...}` and `for (var it = something.iterator; it.moveNext();) { var x = it.current; ...}`, and the former is much easier to use.
We allow `s.isEmpty` and `s.length == 0`, but the former is *much* easier to read.
I see no problem allowing both `[... list]` and `[for (var x in list) x]` and `[for (var i = 0; i < list.length; i++) list[i]]`. You use the one that is easier to read and write, but you have a more general feature to fall back to when you stop being in the optimized case. 

Readers will *expect* you to use the simpler version, and start looking for reasons that you didn't if you use the more general version.

**Comment by tatumizer** [comment]
@lrhn: That's the whole point: I can't imagine any context in which `[for (var x in list) x]` is easier to read, write, meditate about, etc. compared with `[...list]`. What specialized version? To me, the latter is a specialized version. Well, one argument can be made here is that `[for (var i = 0; i < n; i++) f(n)]` does not have a direct equivalent in terms of `...expr`, but the issue of *ranges* was raised several times, there was even a proposal on the table, and eventually dart may add ranges - then we will have 
`[...range(0,n).map(x=>f(x))`]. Which of these two versions is more specialized? Hard to say. If you allow subjectivity here, then the whole argument about "specialized versions" becomes fuzzy, and we are back to exactly 2 equivalent variants doing the same thing, which you are trying to avoid (unless you redefine "specialized version" as "the one I like more tonight") :)

My personal (very subjective) view is that *in the context of dart*, overall, C-style for loop looks a bit antiquated, I would prefer `for ... in ...` in every situation, it's just dart has a couple of holes where `for ... in ...` still can't be used - but the holes can be easily closed IMO. But then, [...expr] will win hands down, in any context. No?

**Comment by pschiffmann** [comment]
> the issue of _ranges_ was raised several times, there was even a proposal on the table, and eventually dart may add ranges - then we will have `[...range(0,n).map(x=>f(x))`].

You can't `continue` or `break` a `.map()` call.

> My personal (very subjective) view is that _in the context of dart_, overall, C-style for loop looks a bit antiquated, I would prefer `for ... in ...` in every situation

In my experience, C-style for loops occupy more of a niche role, but I do have occasional use for it, like
`for (var cur = node; cur != null; cur = current.next)`

**Comment by tatumizer** [comment]
> You can't continue or break a .map() call.

In the context of list/map literals, you cannot continue or break for-element either :). 

> for (var cur = node; cur != null; cur = current.next)

To my taste, "while (cur != null)" loop would be *at least* as good here.

**Comment by lrhn** [comment]
@tatumizer 
There is indeed no situation where `[for (var x in list) x]` is better than `[...list]`, but that is also a very degenerate example of a `for-in` loop. The `for` loop is there for more complex things.
You can also write `[for (var x in list) if (x < 12) x * x]`, for which there is no convenient spread operation.
Sure, you can do `[...list.where((x) => x < 12).map((x) => x * x)]`, but that's not actually easier to read.
Or more efficient, with the two intermediate iterables and function arguments.
Maybe if Dart had a different syntax, like `[... list where: it < 12 map: it * it]`, and proper call-by-name and efficient user-defined control structures, then it would be more useful, but we don't, so it isn't.

We don't want to *prevent* you from writing `[for (var x in list) x]`. 
We could. It would just be a rule of "the element expression of a for-in loop element which declares a loop variable must not be exactly the loop variable". It's just unnecessary noise to add that restriction at the language level. The language gains nothing from making it a hard restrictions. A restriction is an exception, and every exception requires user mental capacity to remember. I'd rather use that for something more important.
I fully expect the analyzer to make a *hint* that you should use a spread instead for that situation.

I also agree that `for-in` loops are more pleasant than the `for(;;)` loops. Again, you can do anything with `for(;;)`, including implement `for-in`, but for the use case of iterating over an iterable sequence of values, where the body cannot interfere with the iteration, `for-in` is the preferred specialized variant.

I dare you to write something like this with `for-in`, though:
```dart
Iterable<int> codePoints(String string) sync* {
  for (int i = 0; i < string.length; i++) {
    var char = string.codeUnitAt(i);
    while (isLeadSurrogate(char) && i + 1 < string.length) {
      var next = string.codeUnitAt(i++);
      if (isTailSurrogate(next)) {
        char = combineSurrogates(char, next);
        break;
      }
      yield char;
      char = next;
    }
    yield char;
  }
}
```
The ability for the body to affect the loop variable(s) can be indispensable. 
When you don't need that, by all means use `for-in`, that's the often-occurring special case that it's there for.

**Comment by tatumizer** [comment]
> Sure, you can do [...list.where((x) => x < 12).map((x) => x * x)], but that's not actually easier to read. 

In Kotlin, it would be simply  `[...list.where{it < 12}.map{it * it}]`, which is not bad. Dart could do something similar and maybe better - not clear what exactly, and *that* is something that deserves a serious look into IMO.
 
Otherwise the logic of your justification for `[for (;;)]` boils down to this: we don't have a concise syntax for lambdas, so let's go with "for-element". But in the next version, we may come up with a better idea for lambdas, and then what? How will we explain the rationale behind for-element then? :).

My point is that for-element (in literals), no matter whether it's for-in or for(;;), is unnecessary. Spread operator [...expr] and better lambdas solve the same problem more nicely. More mathiness please, not less! :)

To be clear, I'm not against `for (;;)` as a necessary low-level construct (After all, I'm an ex-low-level programmer myself, in more than one sense) :-). But you would probably agree that example you provided is a rarity. Statistically, we mostly have the cases you call degenerate - true, they are degenerate in some metaphysical sense, but they are very common practically...

IMO, the language has to provide concise notation for the statistically likely cases and escape hatches (workarounds) for the rare cases. Speaking of lambdas, the most common case of lambda is: single-parameter, single-expression lambda (no blocks). 
In other words, we have to find a symbol for an operator that closurizes the expression. And it better be a single symbol. I think it's a **major** problem for dart, regardless of for loop or literals or anything.
How about this:
```list.where(:it < 12).map(:it * it); // colon as prefix operator, "it" is a magic arg name```

If this is acceptable, I challenge you to provide an example where for-element works better :)

**Comment by tatumizer** [comment]
It can be even shorter! I went out to clean up snow from my driveway, and approx. 30 minutes into the exercise, sudden thought flashed through my mind: **w**. Yes! I recalled very vividly my experiment in compiling a vocabulary of dart language several years back, and one of the major discoveries was that identifier "w" was never used by anybody! I don't know why. Maybe these days, e.g. in Flutter, "w" sometimes denotes Widget, but then, we have "u" - also very rarely used one.  With "u", we could have an absolute beauty here: `list.where(:u<12).map(:u*u)`. 

 In Kotlin, you can't say `var func = {it*it}` - because their syntax for lambdas works only in the context of function invocations (last parameter). In contrast, colon operator can be used anywhere - e.g. it would allow `var func = :u*u`.
I can't imagine anyone not immediately falling in love with this operator.

**Comment by tatumizer** [comment]
In fact, some methods of iterable interface require 2 parameters. They can be called u and w
`list.fold(0, :u+w)`

When compiler doesn't have any context to infer the type of the lambda, it just looks into the body to determine whether w is used. So the specialization works for 1-parameter and 2-parameter lambdas, and nothing else.

I tested performance of `list.where((x)=>x>=0).map((x)=>x*x).toList();`
Compared with equivalent for(;;) or for...in loop, this chain is 3 times slower. In fact, it can be N times slower if the chain contains N hops. Hopefully, something can be done about it.

**Comment by tatumizer** [comment]
@lrhn: given 2 following variants, which one would you prefer, and on what grounds?
Variant 1:
```[for (var x in list) if (x < 12) x * x] ```

Variant 2:
```[...list.where(:u < 12).map(:u * u) ]```

Similar question about for(;;) loop:

Variant 1:
```[for (int i = 0; i < 12; i++) if (someCondition(i)) i * i]```

Variant 2:
```Iterable.generate(12).where(:someCondition(u)).map(:u * u)```

**Comment by lrhn** [comment]
The difference between open and closed iteration (can the loop affect the iteration) is crucial. In most case closed iteration is enough, the body just receives the value. In a few cases, you need open iteration, and then the `for(;;)` is necessary. So, I'd never remove `for(;;)`. It's the same reason we have `sync*` methods: Not all iterables can be written declaratively, some need to be imperative.

The next question is whether you need `for(in)` in a collection literal when you can just do `... iterable` ad create the iterable in the usual ways, instead of generating the individual elements.

I'm not sure which variant I prefer in the first example. "It depends."
They are about equally easy to read. The former has the advantage of not allocating two extra `Iterable` objects, two extra iterator objects and two closures (at least the closures here are simple enough that you only need one in the program, not one per execution).

For example 2, `Iterable.generate(12)` is hard to read, as you say, and it also types badly (it's an `Iterable<dynamic>`). We could add `Iterable.range(0, 12)`, but why stop there?
Let's add a range expression syntax, `[a:b]` is the range from `a` to, but not including, `b`, and you can omit `a` to start at `0`. Then it's the comparison between:
```dart
[for (var it in [:12]) if (someCondition(it)) it * it]
```
vs
```dart
[...[:12]..where(:someCondition(it)).map(:it * it)]
```
Even with the `:` notation, the second variant is longer and I still can't say which is easier to read (I lean towards the former).

For the third example, I think the former variant wins hands down. At least do the mapping on the stream not the list.
```dart
[...await stream.where(:u < 12).map(:u * u).toList()]
```
The `toList()` is still unnecessary overhead that impedes readability and performance.

**Comment by tatumizer** [comment]
@lrhn : let's go one step back: I took a more critical look on your example of codePoints and find it quite difficult to follow. Not sure I understand the algorithm correctly, but here's my shot (for-in instead of for(;;))
```
Iterable<int> codePoints(String string) sync* {
  int prev = null;  
  for (int char in string.codeUnits.followedBy(<int>[null])) {
    if (prev != null && isLeadSurrogate(prev) && char != null && isTailSurrogate(char)) {
      yield combineSurrogates(prev, char);
      prev = null;
    } else {
      yield ?prev; // null-elimination
      prev = char;
    }     
  }
}
```
I think this version is clearer, no?

**Comment by lrhn** [comment]
It's definitely readable.
It relies on a known idiom for iterating pairs (keeping the previous value around until you know what to do with it), so that makes it quickly readable to someone recognizing the idiom, and still decipherable to someone who doesn't.

The setting of `prev` to `null` is really a marker for skipping the next iteration, so you have a pass through the body that does nothing except repopulate the `prev` value.
So, clever! :+1:

**Comment by tatumizer** [comment]
Thank you! Glad you liked it :)

A couple of random thoughts, on "FWIW" basis.

1. I would prefer `range(start, end)` over special notation like `[start:end]`.
The latter looks a bit cryptic and introduces yet another syntactic sugar over a concept that doesn't need more sugar.
Also, `range(start, end)`, IMO, will look more "in style" when used in a chain of functional methods:
```range(0,10).where(:cond(u)).map(:u*u);```
These "where", "map", "fold" etc are all *words*, and "range" is just another word.
If you need a list, you can always write it as `[..range(0,10)]`, which is arguably good enough.
If we could find a word for 1-parameter function xxxx(n) as a synonym to range(0, n), it would be even better, but I was not able to come up with the word that conveys this meaning naturally. I even considered `shovel(n)` (I think it was inspired by my daily routine 
of shoveling snow for the last 2 months), but eventually decided against it (the concept might be too unfamiliar to the users residing in a warmer climate) :)
Anyway, "range" is OK too, AFAIC.

2. Calling a parameter "u" rather than "it" - makes a lot of difference IMO.
I think Kotlin fell into semantic trap with "it", by beleiving "it" can be interpreted as a pronoun, thus making the construct more meaningful.
In fact, in this specific context, it's just a symbol, and has no other meaning but "a parameter" (rather than "this thing", which is meaningless anyway).
As soon as we agree it's just a symbol, why do we need 2 characters where 1 is enough? I got used to "u" within 1 minute and like it a lot :)
Also, we can have both u and w (if we need 2 parameters).

3. I understand the reason why the chain of functional calls like `list.where(...).map(...)` is difficult to optimize: this chain produces an iterable - basically, we have something like a monadic program here. But situation becomes radically different when we write
`[...list.where(...).map(...)]` - here, the whole iterable is *consumed*, and now compiler can inline everything, so performance-wise, it will be indistinguishable from "for-element", 
with much richer vocabulary of operations (for-element doesn't have even a fraction of functionality provided by the chain of methods defined for Iterable).
So it looks that the only remaining justification for for- and if- elements is the inability of compiler, **as of today**, to optimize functional chains.
Is it a correct assessment?

**Comment by lrhn** [comment]
1) I would prefer `range` being available, but adding lower-case top-level declarations to `dart:core` or another often-used `dart:` library is not viable. The scoping rules would mean that `range` the previously meant `this.range` might now refer to the top-level declaration instead, breaking existing code. So, most likely there will be no top-level lower-case declarations in `dart:core` ever again.
Another option is a method on `int`, so `4.range(5)`, but I haven't found a name yet where it's clear whether the argument is inclusive or exclusive. (Too bad integers already have subtraction, otherwise we could make `4 - 8` return `range(4, 8)` [1]. If we ever introduce an infix operator that integers doesn't have, we might get a chance then.)

I wouldn't introduce `[3:8]` as expression syntax without also introducing `operator [:]` and `operator [:]=` as user-definable operators. Then `string[4:8]` or `list[2:]` would be a thing too.

2) Calling a parameter `u` feels arbitrary (because it is), where `it` actually *reads* nicely. Reading `names.forEach(=>print(it))` works. I'd be fine with `_` as well, if we make `_` non-binding as an argument.

3) You can only optimize `[...list.where(test).map(convert)]` if you know the exact type of `list` is a platform list. Otherwise calling `where` or `map` may have side effects, and maybe even the final iteration. If it *is* a platform list, then the compiler can expand it to (effectively) 
```dart
[for (int i = 0; i < list.length; i++) if (test(list[i])) convert(list[i])]
```
It can, and does, do that now for other iterations of platform lists, and recognizing `where` and `map` should be simple enough. If the `test` and `convert` functions are literals, we can even inline them.
It's not new to list literals, we can do the same for a normal iteration:
```dart
for (var x in list.where(test).map(convert)) ...
```
can be compiled to
```
for (var $ in list) if (test($)) { var x = convert($); ... }
```
as well, and the `var $ in list` can be expanded to an index-based iteration.

The problem is that this *only* works for known platform lists (or types where we know the implementation of `where` and `map` for sre), and in ahead-of-time compiled Dart, we can't rely on recognizing that at run-time. Code size can blow up if we do too much speculative case-expansion.

[1]: I had a dream of making an object called `UTF` with a minus operator which returns an encoding depending on the argument value, so `UTF-8` and `UTF-16` would be a thing. :devil:

**Comment by tatumizer** [comment]
> Another option is a method on int, so 4.range(5), but I haven't found a name yet where it's clear whether the argument is inclusive or exclusive  

I've just found a name! 0.until(10). I found it in Kotlin, of all places (the word is good, but the syntax Kotlin uses is ugly). Kotlin also has upTo and downTo (inclusive) and steps - IMO, this is unnecessary: "until" covers 99%  of cases (as usual in statistics, the number is made-up), and that's good enough. YES?

> Calling a parameter u feels arbitrary (because it is), where it actually reads nicely. Reading names.forEach(=>print(it)) works...

Fine, let it be "it", I missed another possible interpretation of "it" as "value for the next **it**eration", so the "superposition" of 2 meanings of "it" (the first being simply a pronoun) is cute - so I take my words back.
I just tried to somehow find a way to support 2-parameter calls like fold, so "u" and "w" seemed like a good idea, but probably it is not. For "fold", we can explicitly write it as `list.fold((val, it)=>...)` - thus preserving the meaning of "it" without relying on built-in shortcuts. 

> It's not new to list literals, we can do the same for a normal iteration: `for (var x in list.where(test).map(convert)) ...`

Yes, it dawned on me later, too :)

> The problem is that this only works for known platform lists

Which is 99.99% of cases :). After substitution, we get "The problem is that this only works in 99.99% of cases". I think it would be tolerable, no? :)

I started an investigation of java's performance on streams (that's how java calls dart's Iterable).
The effort is just in its initial stages, I will post an update when it becomes available, but not anytime soon, b/c the operation is very labor-intensive, and we are expecting another snowfall shortly, which will keep me busy for the next 1 or 2 weeks. Hopefully, we will see better days around July  :)

**Comment by ds84182** [comment]
Perhaps two parameter calls can be supported with `it` and `it1`, e.g. `[1, 2, 3, 4].reduce(=> it + it1)`?

**Comment by tatumizer** [comment]
As expected, java optimizes functional calls, and doing a very good job at it.
The test:
java: ```list.stream().filter(x->x>=0).map(x->x*x).collect(Collectors.toList());```
dart: ```list.where((x)=>x>=0).map((x)=>x*x).toList();```

Because the timing depends on the list size, the program is trying to approximate it as
t(n) = a + b\*n, where n - list size, a - constant overhead (related to allocation and GC), b - cost of a single element (a and b are measured in clock cycles =timeInNanoseconds*processorClockRate).

I'm running both java and dart with 512m of memory.
There are 4 runs in total: short warmup with n=16, then program runs for about 1 sec with each n=16,n=32 and n=64 where n=16 and n=64 are used to solve the system of equations for a and b, n=32 - to check the error in interpolation (just to be sure the things are not terribly non-linear). 

For dart, t(n)= 260 + 90\*n
For java, t(n)= 425 + 37\*n (see \*)

For reference, when written as a "for" loop, the timing is:
For dart: t(n) = 80 + 28\*n
For java, t(n) = 127+24\*n

When I add one extra step map((u) - > u+1), java timing increases by just 7 cycles, whereas dart ... let's not talk about it :)

Conclusion: if java wanted to implement the feature currently targeted by if-element and for-element, it won't need anything beyond spread operator - at least as far as performance is concerned.
 
(\*) I know it's not exactly apples to apples, b/c in fact, the timing generally is not O(n) but rather O(n*log n) due to reallocations; java allocates the list with just 10 elements initially; for dart - I don't know the number, probably more, and that appears to affect both coefficients -  but anyway, the qualitative picture is clear.

**Comment by tatumizer** [comment]
Update: if someone is still interested: in java, if you use an int array instead of a list as a source, the performance is almost identical to that of manually written "for" loop. 
Tested on expression 
```Arrays.stream(array).filter(u -> u >= 0).map(u -> u * u).toArray();```
